crates/ruff_python_formatter/src/expression/expr_tuple.rs

```diff
@@ -19,14 +19,17 @@ pub enum TupleParentheses {
     Default,
     /// Effectively `Some(Parentheses)` in `Option<Parentheses>`
     Expr(Parentheses),
+    /// TODO(cnpryer): Why not TupleParentheses::Expr(Parentheses::Never)?
+    Never,
```

Comment:
Could you describe the use case where this is required, ideally with a python code snippet and how black formats this differently from the other variants?

Reply:
Sure. I'll compile some examples tomorrow.

Reply:
So the first example that motivated this was #5790. 

```py
# Input
{k: v for k, v in this_is_a_very_long_variable_which_will_cause_a_trailing_comma_which_breaks_the_comprehension}

# black
{
    k: v
    for k, v in this_is_a_very_long_variable_which_will_cause_a_trailing_comma_which_breaks_the_comprehension
}
```
`black` will *never* format tuple `for` targets with parentheses if the target tuple is found in a comprehension. 

> how black formats this differently from the other variants

If I understand this, in other variations of statements or expressions tuples often don't format this way. For example

```py
# Input
for a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a in []:
    ...

# Black
for (
    a,
    a,
    a,
    # ...
    a,
    a,
) in []:
    ...
```
This for loop is not found in a comprehension and formats with expended multi-line wrapping parentheses.

For the comprehension implementation I understood its `TupleParentheses` options to be "here we *never* want parentheses.

A similar motivation came from #5806.

> It would be great if we could reduce this enum to as few variants as possible to reduce complexity

I have new thoughts here. 
1. I feel like *complexity* can be defined as "many variants" but it could also be more *complex* to try to build special case formatting with *less* variants. I'm leaning towards less variants anyway, but I thought I'd share that. 
1. If we want less variant surface area for `TupleParentheses`, why would we stop there? I haven't encountered this yet contributing here, but would it be better to have a broader enum like `SpecialParentheses` for more than just special tuple cases? (terrible name but hopefully you get the point)

The last thing I wanted to bring up was from #5708 
> \- `maybe_parenthesize_expression` is a new builder that may parenthesize an expression depending on `Parenthesized` and the expressions `NeedsParentheses` implementation.

`maybe_parenthesize_expression` clicked immediately for me when I first used it. Could we not use a combination of `Parentheses` and `NeedsExpression` to manage all cases? If we expand `Parentheses` for more than just general parenthesize and use it as fundamental building blocks, couldn't we tell `maybe_parenthesize_expression` what to do using variants from an expanded `Parentheses`?

_____

Does my motivation for replacing `Comprehension` and `StripInsideForLoop` at least make sense? I'm just trying to gauge if I'm heading down the right path. It sounds like at least this PR in some form is wanted.

Reply:
Re-read [this](https://github.com/astral-sh/ruff/pull/5790#discussion_r1267598124). My second thought above (2) doesn't seem like a good path at this time.

Reply:
I added this example to 2e1ab30efd436370c2e6292743a65941662c106a

Reply:
Thanks for example. It seems like you've identified the *minimal* set, because each variant differs in at least one property from the other variants.

Reply:
Marking this as resolved